src/RMPI/mpidata.jl

```diff
@@ -233,64 +233,84 @@ function LinearAlgebra.dot(md::MPIData, lop, x)
     return MPI.Allreduce(dot(localpart(md), lop, x), +, md.comm)
 end
 
-function LinearAlgebra.dot(md_left::MPIData, lop, md_right::MPIData)
-    # Strategy: accumulate the dot product without allocating an intermediate vector.
-
-    # spawns to other ranks are stored in buffers and communicated via MPI point-point
-    # communications.
+"""
+    copy_to_local(md::MPIData)
 
-    T = promote_type(eltype(lop), valtype(md_left), valtype(md_right))
-    dv = localpart(md_left)
-    comm = md_left.comm
+Collect all pairs in `md` from all ranks and store them in a local `AbstractDVec`.
+"""
+copy_to_local(md::MPIData) = copy_to_local!(similar(localpart(md)), md)
 
-    P = Pair{keytype(md_right),T}
-    buffers = Vector{P}[P[] for _ in 1:mpi_size(comm)]
-    myrank = mpi_rank()
+"""
+    copy_to_local!(target, md::MPIData)
 
-    result = zero(T)
+Collect all pairs in `md` from all ranks and store them in `target`.
+In-place version of [`copy_to_local`](@ref).
+"""
+function copy_to_local!(target, md::MPIData)
+    copy!(target, localpart(md))
+    comm = md.comm
+    myrank = mpi_rank(comm)
+    datatype = MPI.Datatype(eltype(md))
 
-    # Sort values into buffers and communicate.
-    for (key, val) in pairs(localpart(md_right))
-        result += conj(dv[key]) * diagonal_element(lop, key) * val
-        for (add, elem) in offdiagonals(lop, key)
-            tr = targetrank(add, mpi_size(comm))
-            if tr == myrank
-                result += conj(dv[add]) * elem * val
-            else
-                push!(buffers[tr + 1], add => elem * val)
-            end
-        end
+    # Store all pairs to a buffer.
+    sendbuf = Vector{eltype(md)}(undef, length(localpart(md)))
+    for (i, p) in enumerate(pairs(localpart(md)))
+        sendbuf[i] = p
     end
```

Comment:
Isn't this the same as the more concise
```suggestion
    # Store all pairs to a buffer.
    sendbuf = collect(pairs(localpart(md)))
```
?

Reply:
It is.